File: backend/app/schemas/inventario_animal.py
from pydantic import BaseModel, Field
from typing import Optional
from datetime import date
from enum import Enum
from app.schemas.ubicacion import UbicacionOut
from app.schemas.proveedor import ProveedorOut
from app.schemas.lote import LoteBase
from datetime import datetime

class EstadoAnimal(str, Enum):
    Activo = "Activo"
    Vendido = "Vendido"
    Muerto = "Muerto"
    Descartado = "Descartado"

class EspecieOut(BaseModel):
    especie_id: int
    nombre_comun: str

    class Config:
        from_attributes = True

class RazaOut(BaseModel):
    raza_id: int
    nombre_raza: str

    class Config:
        from_attributes = True

class LoteOut(LoteBase):
    lote_id: int
    fecha_creacion: date

    class Config:
        from_attributes = True

class MotivoIngreso(str, Enum):
    Nacimiento = "Nacimiento"
    Compra = "Compra"
    TrasladoInterno = "TrasladoInterno"

class MotivoEgreso(str, Enum):
    Venta = "Venta"
    Muerte = "Muerte"
    Descarte = "Descarte"
    TrasladoExterno = "TrasladoExterno"
    

class AnimalSimpleOut(BaseModel):
    animal_id: int
    numero_trazabilidad: str
    nombre_identificatorio: Optional[str] = None
    especie: EspecieOut
    raza: RazaOut
    sexo: str
    fecha_nacimiento: Optional[date] = None
    estado_actual: EstadoAnimal
    fecha_registro: datetime
    observaciones_generales: Optional[str] = None

    class Config:
        from_attributes = True

class InventarioAnimalBase(BaseModel):
    animal_id: int
    fecha_ingreso: date
    motivo_ingreso: MotivoIngreso
    proveedor_compra_id: Optional[int] = None
    precio_compra: Optional[float] = None
    ubicacion_actual_id: Optional[int] = None
    lote_actual_id: Optional[int] = None
    fecha_egreso: Optional[date] = None
    motivo_egreso: Optional[MotivoEgreso] = None

class InventarioAnimalCreate(InventarioAnimalBase):
    pass

class InventarioAnimalUpdate(InventarioAnimalBase):
    activo_en_finca: Optional[bool] = None # <-- ¡Aquí está el cambio!

class InventarioAnimalOut(InventarioAnimalBase):
    inventario_id: int
    activo_en_finca: bool
    animal: AnimalSimpleOut
    ubicacion_actual: Optional[UbicacionOut] = None
    lote_actual: Optional[LoteOut] = None
    proveedor_compra: Optional[ProveedorOut] = None  # ✅ correcto nombre
    

    class Config:
        from_attributes = True

File: types/inventarioAnimal.ts
// types/inventarioAnimal.ts

import { UbicacionOut } from "@/hooks/useUbicaciones";
import { AnimalOut } from "./animal";
import { LoteOut } from "@/hooks/useLotes";
import { ProveedorOut } from "./proveedor";

export enum MotivoIngreso {
  Nacimiento = "Nacimiento",
  Compra = "Compra",
  TrasladoInterno = "TrasladoInterno",
}

export enum MotivoEgreso {
  Venta = "Venta",
  Muerte = "Muerte",
  Descarte = "Descartado",
  TrasladoExterno = "TrasladoExterno",
}

export interface InventarioAnimalBase {
  animal_id: number;
  fecha_ingreso: string; // Cambiado a string para coincidir con la fecha en formato ISO (YYYY-MM-DD)
  motivo_ingreso: MotivoIngreso;
  proveedor_compra_id?: number | null;
  precio_compra?: number | null; // Cambiado a number | null para representar Optional[float]
  ubicacion_actual_id?: number | null;
  lote_actual_id?: number | null;
  fecha_egreso?: string | null; // Cambiado a string | null para coincidir con la fecha en formato ISO (YYYY-MM-DD)
  motivo_egreso?: MotivoEgreso | null;
}

export interface InventarioAnimalCreate extends InventarioAnimalBase {}

export interface InventarioAnimalUpdate extends InventarioAnimalBase {}

export interface InventarioAnimalOut extends InventarioAnimalBase {
  inventario_id: number;
  activo_en_finca: boolean;
  animal: AnimalOut;
  ubicacion_actual?: UbicacionOut | null;
  lote_actual?: LoteOut | null;
  proveedor_compra?: ProveedorOut | null;
}
